modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js

```diff
@@ -174,26 +178,72 @@ export default class VectorTileFeature {
   }
 }
 
-/**
- * In the original GeoJSON implementation this function
- * classifies the rings in a polygon into outlines & holes,
- * based on the vertex winding order. However the implementation
- * was broken and all it did was classify each ring as a separate
- * polygon outline. In order to maintain compatibility with the
- * original, we keep this behavior, but simplify the implementation
- * to an equivalent function:
- * [A, B, C] => [[A], [B], [C]]
- *
- * TODO in the future, the classification could be fixed following
- * https://github.com/mapbox/vector-tile-js, but this would require
- * confirming that the renderer of deck.gl is actually capable of
- * drawing polygons with holes in them.
- */
-function classifyRings(rings) {
-  return rings.map(r => [r]);
+// classifies an array of rings into polygons with outer rings and holes
+function classifyRings(geom) {
+  const len = geom.lines.length;
+
+  if (len <= 1) return [geom.lines];
+
+  const polygons = [];
+  let polygon;
+  let ccw;
+  let offset = 0;
+
+  for (let i = 0, startIndex, endIndex; i < len; i++) {
+    startIndex = geom.lines[i] - offset;
+
+    endIndex = geom.lines[i + 1] - offset || geom.data.length;
+    const shape = geom.data.slice(startIndex, endIndex);
+    const area = signedArea(shape);
+
+    if (area === 0) {
+      // This polygon has no area, so remove it from the shape
+      // Remove the section from the data array
+      const before = geom.data.slice(0, startIndex);
+      const after = geom.data.slice(endIndex);
+      geom.data = before.concat(after);
+
+      // Need to offset any remaining indices as we have
+      // modified the data buffer
+      offset += endIndex - startIndex;
+
+      // Do not add this index to the output and process next shape
+      continue;
+    }
+
+    if (ccw === undefined) ccw = area < 0;
+
+    if (ccw === area < 0) {
+      if (polygon) polygons.push(polygon);
+      polygon = [startIndex];
+    } else {
+      // @ts-ignore
+      polygon.push(startIndex);
+    }
+  }
+  if (polygon) polygons.push(polygon);
+
+  return polygons;
+}
+
+function signedArea(ring) {
```

Comment:
Nit: Seems like functions we have in `@math.gl/polygon` but maybe this implementation is MVT format specific?

Reply:
Seems identical to `getPolygonSignedAreaPoints`?

https://github.com/uber-web/math.gl/blob/cead84f7a7a7c068ad85b9cb245abe4655009e47/modules/polygon/src/polygon-utils.js#L107-L118

Reply:
You are correct. I looked at the implementation in @math.gl and although I was a bit worried about the number of function invocations compared to this implementation, I did some perf testing and there is no measurable difference. I'll switch to using the implementation from @math.gl